groups/nts/ntsu/ntsu_socketutil.cpp

```diff
@@ -539,34 +542,180 @@ void IncomingDataStats::update(const struct iovec* buffersReceivable,
+    };
+
+    // Define a type alias for a maximimally-aligned buffer of suitable size to
+    // send any meta-data (viz. open file descriptors) to the peer of the 
+    // socket.
+    typedef bsls::AlignedBuffer<k_SEND_CONTROL_BUFFER_SIZE> Arena;
+
+    Arena d_arena;
+
+  private:
+    SendControl(const SendControl&) BSLS_KEYWORD_DELETED;
+    SendControl& operator=(const SendControl) BSLS_KEYWORD_DELETED;
+
+  public:
+    // Create a new send control buffer.
+    SendControl();
+        
+    // Destroy this object.
+    ~SendControl();
+        
+    // Encode the specified 'options' into the control buffer. Return the 
+    // error.
+    ntsa::Error encode(msghdr* msg, const ntsa::SendOptions& options);
+};
+
+// Provide a buffer suitable for attaching to a call to 'recvmsg' to receive
+// control data from the peer of a socket.
+class ReceiveControl 
+{
+    enum {
+        // The payload size required to receive any meta-data  (e.g. open 
+        // file descriptors, timestamps, etc.) buffered by the operating system
+        // for a socket.
+        k_RECEIVE_CONTROL_PAYLOAD_SIZE = static_cast<int>(
+            NTSU_SOCKETUTIL_MAX_HANDLES_PER_INCOMING_CONTROLMSG * sizeof(int))
 #if defined(BSLS_PLATFORM_OS_LINUX)
-namespace {
+            + static_cast<int>(sizeof(TimestampUtil::ScmTimestamping))
+#endif
+
+        // The control buffer capacity required to receive any meta-data (e.g.
+        // open file descriptors, timestamps, etc.) buffered by the operating
+        // system for a socket.
+        , k_RECEIVE_CONTROL_BUFFER_SIZE = 
+            static_cast<int>(CMSG_SPACE(k_RECEIVE_CONTROL_PAYLOAD_SIZE))
+    };
+
+    // Define a type alias for a maximimally-aligned buffer of suitable size to
+    // receive any meta-data (e.g. foreign handles, timestamps, etc.) buffered
+    // by the operating system for a socket.
+    typedef bsls::AlignedBuffer<k_RECEIVE_CONTROL_BUFFER_SIZE> Arena;
+
+    Arena d_arena;
+
+  private:
+    ReceiveControl(const ReceiveControl&) BSLS_KEYWORD_DELETED;
+    ReceiveControl& operator=(const ReceiveControl) BSLS_KEYWORD_DELETED;
+
+  public:
+    // Create a new receive control buffer.
+    ReceiveControl();
+        
+    // Destroy this object.
+    ~ReceiveControl();
+
+    // Zero the control buffer.
+    void initialize(msghdr* msg);
+        
+    // Decode the control buffer of specified 'msg' and decode its contents 
+    // into the specified 'context' according to the specified 'options'.
+    ntsa::Error decode(ntsa::ReceiveContext*       context, 
+                       const msghdr&               msg,
+                       const ntsa::ReceiveOptions& options);
+};
+
+NTSCFG_INLINE
+SendControl::SendControl()
+{
+}
+
+NTSCFG_INLINE
+SendControl::~SendControl()
+{
+}
+
+ntsa::Error SendControl::encode(msghdr* msg, const ntsa::SendOptions& options)
+{
+    bsl::memset(d_arena.buffer(), 0, k_SEND_CONTROL_BUFFER_SIZE);
```

Comment:
I think it makes sence to put this `memset` line after the `if (options.foreignHandle().isNull())` block. What's the point in zeroing the memory if we have nothing to encode?

Reply:
Done.